crates/noirc_evaluator/src/ssa_refactor/ir/function_inserter.rs

```diff
@@ -0,0 +1,129 @@
+use std::collections::HashMap;
+
+use iter_extended::vecmap;
+
+use super::{
+    basic_block::BasicBlockId,
+    dfg::InsertInstructionResult,
+    function::Function,
+    instruction::{Instruction, InstructionId},
+    value::ValueId,
+};
+
+/// The FunctionInserter can be used to help modify existing Functions
+/// and map old values to new values after re-inserting optimized versions
+/// of old instructions.
+pub(crate) struct FunctionInserter<'f> {
+    pub(crate) function: &'f mut Function,
+
+    values: HashMap<ValueId, ValueId>,
+}
+
+impl<'f> FunctionInserter<'f> {
+    pub(crate) fn new(function: &'f mut Function) -> FunctionInserter<'f> {
+        Self { function, values: HashMap::new() }
+    }
+
+    /// Resolves a ValueId to its new, updated value.
+    /// If there is no updated value for this id, this returns the same
+    /// ValueId that was passed in.
+    pub(crate) fn resolve(&mut self, value: ValueId) -> ValueId {
+        match self.values.get(&value) {
+            Some(value) => *value,
+            None => match &self.function.dfg[self.function.dfg.resolve(value)] {
+                super::value::Value::Array { array, element_type } => {
+                    let array = array.clone();
+                    let element_type = element_type.clone();
+                    let new_array = array.iter().map(|id| self.resolve(*id)).collect();
+                    let new_id = self.function.dfg.make_array(new_array, element_type);
+                    self.values.insert(value, new_id);
+                    new_id
+                }
+                _ => value,
+            },
```

Comment:
This is where the main fix of this PR. The rest is a mix of refactorings so that I can use a FunctionInserter instead of copying this fix everywhere, and added `.clone()` calls.